rfc: deterministic proto bytes serialization (#7427)
tendermint/tendermint: docs/rfc/rfc-007-deterministic-proto-bytes.md

# RFC 008 : Deterministic Proto Byte Serialization

## Changelog

- 09-Dec-2021: Initial draft (@williambanfield).

## Abstract

This document discusses the issue of stable byte-representation of serialized messages
within Tendermint and describes a few possible routes that could be taken to address it.

## Background

We use the byte representations of wire-format proto messages to produce
and verify hashes of data within the Tendermint codebase as well as for
producing and verifying cryptographic signatures over these signed bytes.

The protocol buffer [encoding spec][proto-spec-encoding] does not guarantee that the byte representation
of a protocol buffer message will be the same between two calls to an encoder.

> While there is a mode to force the encoder to produce the same byte representation

Author: Yeah, it's written down in the spec, I know it exists in c++ as well.

https://github.com/protocolbuffers/protobuf/releases/tag/v3.0.0

1. Encoding order of fields within a message.

Proto allows fields to be encoded in any order and even be repeated.

2. Encoding order of elements of a repeated field.

`repeated` fields in a proto message can be serialized in any order.

3. Presence or absence of default values.

Types in proto have defined default values similar to Go's zero values. 
Writing or omitting a default value are both legal ways of encoding a wire message.

4. Serialization of 'unknown' fields. 

Unknown fields can be present when a message is created by a binary with a newer 
version of the proto that contains fields that the deserializer in a different 
binary does not yet know about. Deserializers in binaries that do not know about the field
will maintain the bytes of the unknown field but not place them into the deserialized structure.

We have a few options to consider when producing this stable representation.

### Use only compliant serializers and constrain field usage

According to [Cosmos-SDK ADR-27][cosmos-sdk-adr-27], when message types obey a simple 
set of rules, gogoproto produces a consistent byte representation of serialized messages.

> This seems promising, although more research is needed to guarantee gogoproto always

Reviewer: Don't we also have the intention of removing gogoproto?

Author: I believe that there is interest in doing this in the SDK, although I'm not sure I fully understand the timeline. It's likely a reasonable thing to do seeing as gogoproto is currently in maintenance limbo. However, in terms of concrete plans to remove it from Tendermint completely, I'm not aware of any at the moment.

produces a consistent set of bytes on serialized messages. This would solve the problem 
within Tendermint as written in Go, but would require ensuring that there are similar

> serializers written in other languages that produce the same output as gogoproto.

Author: I think a test harness that encodes the structure as JSON or YAML and then also includes the expected output of marshaling and bit-munging may be worth producing. This would allow projects in different languages to check their implementations against our own.

### Reorder serialized bytes to ensure determinism.

The serialized form of a proto message can be transformed into a canonical representation
by applying simple rules to the serialized bytes. Re-ordering the serialized bytes
would allow Tendermint to produce a canonical byte representation without having to 
simultaneously maintain a custom proto marshaller. 

This could be implemented as a function in many languages that performed the following steps:

1. Reordered all fields to be in tag-sorted order.

> 2. Reordered all `repeated` sub-fields to be in lexicographically sorted order.

Reviewer 2: Q: Is this already the case? I know that for instance ValidatorSets are sorted already but is e.g. is Evidence lexicographically sorted?

Reviewer 3: There’s no guarantee that sorting the field before encoding affects the encoding. In Go it does, and in practice it’s a sensible assumption. But it could change at any time.

3. Deleted all default values from the representation.
4. Set our proto decoder package to remove unknown fields on deserialization.

This would still require that messages never unmarshal data structures with unknown fields.

> This can be accomplished by defining adding fields to a structure that needs canonicalization

Reviewer: ```suggestion
This can be accomplished by defining the act of adding fields to a structure that needs canonicalization
```

> as a breaking changes within our semantic versioning and disallowing it within a major version.

Reviewer: ```suggestion
as a breaking change within our semantic versioning and disallowing it within a major version.
```

I.e., two versions of Tendermint that have different fields in proto messages which require
canonical byte representation will not be able to interoperate.

A prototype implementation by @creachadair of this can be found in [the wirepb repo][wire-pb].
This could be implemented in multiple languages more simply than ensuring that there are

> canonical proto serializers that match in each language.

Reviewer 2: It remains a bit vague what wirepb actually is. Is it some kind of linter that enforces the above rules? Or does it take the raw bytes of any proto serializer and additionally canonicalizes them (by applying above rules)? By quickly skimming through that repo it seems to be the latter and the introductory paragraph also speaks of "re-ordering the serialized bytes". I think it could still be clearer because above steps seem to operate on fields (which first sounds like this is about pre-serialization.

Independent from the text: This is an interesting approach and I like it :)

Reviewer 3: > It remains a bit vague what wirepb actually is.

Right now it is just a quick proof-of-concept. I just added some more documentation to the README, that I hope may help clarify. In summary: It is not just a linter, it actually permutes the content of the wire message in binary format. It operates only on the wire format, and does not know anything about the language-specific type structure or the descriptor schema.

However, it does have some important limitations, which I've spelled out explicitly in the README file.

Another approach, that is potentially more robust, is to make the caller provide a protobuf `Descriptor` message for the expected type along with the byte string to be hashed. Given the descriptor, the limitations of the prototype could be eliminated. However, this would mean every tool that wanted to check the hash of a protobuf message would need to (a) know the expected protobuf type, (b) have access to the descriptor table, and (c) have access to a descriptor-aware implementation of hashing.

Protobuf wire format is simple enough that it's easy to re-implement this algorithm in any language. Adding in descriptor support would make it a much bigger job. It could maybe be worthwhile, if the alternative is having to maintain a completely-parallel deterministic code generator, but there are a lot of unknowns.

In Go specifically, it might be possible to shim in a deterministic marshaling rule via the [Message interface](https://pkg.go.dev/google.golang.org/protobuf@v1.27.1/reflect/protoreflect#Message), but it's not at all clear how we could keep such an implementation synchronized with other languages (and I know at least Rust and JS potentially need to harmonize with the Go implementation—perhaps there are others).

Author: The more complex the implementation, the more difficult it will be to make it agree across languages in all cases. I would prefer that we stick with a simpler implementation. 

Per the limitations in that document:

> The algorithm cannot distinguish between a field of opaque string type that contains the wire encoding of a message, and a field of message type.

This is worth keeping in mind if we ever allow opaque strings to represent messages in our types. The obvious example that comes to mind is vote-extensions. If we allow an extension to appear as bytes within our structure, we risk mangling the message if we re-encode it. I think being clear that this is strictly for hashing is a fine restriction to put on this algorithm. I.e., do not use this as any kind of intermediate representation ever. Just for comparison of hashes and signatures. 

> The algorithm does not unify default-valued fields with unset (omitted) fields. In proto3, the only default values are equivalent to omitting the field from the message, and most encoders do omit them. However, if an encoder does include default-valued fields in its output, Algorithm 2 will generate a different string than if the field was omitted.

I'll do some minor research to determine which encoders produce which type of output. While we are not guaranteed that this will change in the future, we can at least guide users towards 'safe' encoders in their language and caution users in new languages about what behavior their encoder must adhere to.

> The algorithm does not handle "packed" repeated fields of scalar type. Packed repeated fields are encoded as a wire-type string of concatenated varint values. The values within the string do not have the structure of a message, so the algorithm does not "fix" the order of the packed values.

It appears from my research that when we use a message containing a list, we implement an additional `hash` method on top of it that loops over the list in a deterministic order, marshals the list elements into byte slices, and then merkelizes the resulting slices, so we are generally in the clear on this point so far and can continue to use the same trick in the future.

Reviewer 3: > > The algorithm cannot distinguish between a field of opaque string type that contains the wire encoding of a message, and a field of message type.
> 
> This is worth keeping in mind if we ever allow opaque strings to represent messages in our types.

We need to account for it even if we don't do that: The workaround is to assume all fields with the string wire-type _could_ be messages. So an opaque string that happens to be a syntactically valid wire-format message would also be permuted, even if we never intended to treat it as a message.

We're still hashing all the same bytes in that case—but in a different order from the original. Simple cases (e.g., "\n\tabcdefghi" is one example) would not be changed, but more complex ones would (e.g., "\n\tbcdefghij\n\tabcdefghi" would be come "\n\tabcdefghi\n\tbcdefghij"). As a rule, most strings won't look syntactically like messages, but we can't hang anything important on that.

> I'll do some minor research to determine which encoders produce which type of output. 

It would also be possible to filter such fields out, but as far as I know the proto3 encoders should already be doing this.

> It appears from my research that when we use a message containing a list, we implement an additional `hash` method on top of it that loops over the list in a deterministic order

Fortunately, field packing only applies to scalar numeric types, so repeated messages, strings, etc., are not affected.

Author: > We need to account for it even if we don't do that: The workaround is to assume all fields with the string wire-type could be messages.

To clarify: I more meant that we need to take it into account so as to not accidentally ever deserialize _out_ of the bytes being used for hashing. The canonicalizer needs to always act as though it's acting on a sub-message since it can't disambiguate.

Author: > From my quick investigation, it looks like `protobufjs` does include the serialized bytes on default vaules:

https://github.com/williambanfield/proto-research/blob/master/js/test.js#L26 
https://github.com/williambanfield/proto-research/blob/master/js/output

Finally, we should add clear documentation to the Tendermint codebase every time we

> compare hashes of proto messages or use proto serialized bytes to produces a

Reviewer: ```suggestion
compare hashes of proto messages or use proto serialized bytes to produce
```

digital signatures that we have been careful to ensure that the hashes are performed

> properly.

Reviewer 2: 👍🏼
